src/Analysis/Engine/Impl/ProjectEntry.cs

```diff
@@ -46,7 +46,7 @@ internal sealed class ProjectEntry : IPythonProjectEntry, IAggregateableProjectE
         private readonly HashSet<AggregateProjectEntry> _aggregates = new HashSet<AggregateProjectEntry>();
 
         private TaskCompletionSource<IModuleAnalysis> _analysisTcs = new TaskCompletionSource<IModuleAnalysis>();
-        private AnalysisUnit _unit;
+        internal AnalysisUnit _unit;
```

Comment:
Why do you need it to be `internal`?

Reply:
It's used in ModuleInfo.AnalysisUnit.

Reply:
With some exceptions, fields should be private, they should be wrapped with properties or access methods. But in this case, it is not safe to provide `AnalysisUnit` from `ProjectEntry` for public access. It may be changed while analysis is in progress and that would create unpredictable results.

Reply:
I could add a `AnalysisUnit` to `ModuleInfo` in a similar way it exists in `ClassInfo`. Then `ProjectEntry` will have control over when to change it. However, if you are talking about potentially multithreaded analysis, they can still be out of sync.

Reply:
Analysis is single threaded, but it is in parallel with parsing. Unfortunately, you can't do it like with `ClassInfo`, because there is only one instance of `ModuleInfo` per `ProjectEntry` - it isn't recreated when new analysis arrives.

Reply:
Is there a reason for `ClassInfo`s to be recreated and and `ModuleInfo`s not?

Reply:
It looks like `_unit` reassignment happens just before clearing `ModuleInfo` contents anyway.

Reply:
> Is there a reason for `ClassInfo` to be recreated and `ModuleInfo` not?

`ModuleReference` (https://github.com/Microsoft/python-language-server/blob/2935897bfd8244dbe4296101a80021400ff3c88e/src/Analysis/Engine/Impl/ModuleReference.cs#L37) and some other parts of module dependency analysis work through instance references.

Reply:
I am open for suggestions. Actually, at this moment I don't understand how can parsing be completed in parallel to analysis without race conditions. Here's the relevant code excerpt from `ProjectEntry.Parse`:

```csharp
_unit = new AnalysisUnit(tree, MyScope.Scope);
AnalysisLog.NewUnit(_unit);

MyScope.Scope.Children = new List<InterpreterScope>();
MyScope.Scope.ClearNodeScopes();
MyScope.Scope.ClearNodeValues();
MyScope.Scope.ClearLinkedVariables();
MyScope.Scope.ClearVariables();
MyScope.ClearReferencedModules();
MyScope.ClearUnresolvedModules();
_unit.State.ClearDiagnostics(this);
```

There does not seem to be anything, that would prevent modification of MyScope.* if there might be an analysis of it going on at the same time.